sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c

```diff
@@ -0,0 +1,410 @@
+              GNRC_IPV6_NIB_NC_INFO_AR_STATE_GC);
+}
+
+_nib_t *_nib_nc_add(const ipv6_addr_t *addr, unsigned iface, uint16_t cstate)
+{
+    assert(cstate != GNRC_IPV6_NIB_NC_INFO_NUD_STATE_DELAY);
+    assert(cstate != GNRC_IPV6_NIB_NC_INFO_NUD_STATE_PROBE);
+    assert(cstate != GNRC_IPV6_NIB_NC_INFO_NUD_STATE_REACHABLE);
+    _nib_t *nib = _nib_alloc(addr, iface);
+    if (nib != NULL) {
+        DEBUG("nib: Adding to neighbor cache (addr = %s, iface = %u)\n",
+              ipv6_addr_to_str(addr_str, addr, sizeof(addr_str)), iface);
+        if (!(nib->mode & _NC)) {
+            nib->info &= ~GNRC_IPV6_NIB_NC_INFO_NUD_STATE_MASK;
+            nib->info |= (cstate & GNRC_IPV6_NIB_NC_INFO_NUD_STATE_MASK);
+            nib->mode |= _NC;
+        }
+        if (nib->next == NULL) {
+            DEBUG("nib: queueing (addr = %s, iface = %u) for potential removal\n",
+                  ipv6_addr_to_str(addr_str, addr, sizeof(addr_str)), iface);
+            /* add to next removable list, if not already in it */
+            clist_rpush(&_next_removable, (clist_node_t *)nib);
+        }
+    }
+    else {
+        _nib_t *first = (_nib_t *)clist_lpop(&_next_removable);
+        _nib_t *tmp = first;
+
+        DEBUG("nib: Searching for replaceable entries (addr = %s, iface = %u)\n",
+              ipv6_addr_to_str(addr_str, addr, sizeof(addr_str)), iface);
+        if (tmp != NULL) {
+            do {
+                if (_is_gc(tmp)) {
+                    DEBUG("nib: Removing neighbor cache entry (addr = %s, "
+                          "iface = %u) ",
+                          ipv6_addr_to_str(addr_str, &tmp->ipv6,
+                                           sizeof(addr_str)),
+                          _nib_get_if(tmp));
+                    DEBUG("for (addr = %s, iface = %u)\n",
+                          ipv6_addr_to_str(addr_str, addr, sizeof(addr_str)),
+                          iface);
+                    nib = tmp;
+                    nib->info = 0;
+                    nib->mode = _EMPTY;
+                    _override_node(addr, iface, nib);
+                    nib->info |= (cstate & GNRC_IPV6_NIB_NC_INFO_NUD_STATE_MASK);
+                    nib->mode |= _NC;
+                    break;
+                }
+                else {
+                    /* requeue if not garbage collectible at the moment */
+                    DEBUG("nib: Requeing (addr = %s, iface = %u)\n",
+                          ipv6_addr_to_str(addr_str, &tmp->ipv6,
+                                           sizeof(addr_str)),
+                          _nib_get_if(tmp));
+                    clist_rpush(&_next_removable, (clist_node_t *)tmp);
+                }
+                tmp = (_nib_t *)clist_lpop(&_next_removable);
+            } while (tmp != first);
+        }
+    }
+    return nib;
+}
+
+_nib_t *_nib_iter(const _nib_t *last)
+{
+    bool found = false;
+
+    for (unsigned i = 0; i < GNRC_IPV6_NIB_NUMOF; i++) {
```

Comment:
But then non-empty nodes are skipped. This may be valid if e.g. if a node is already in neighbor cache, but a default router with the same address is added (resulting in `node->mode` to be `_NC | _DRL`).